pkg/policy/policy_controller.go

```diff
@@ -566,6 +570,49 @@ func getTrigger(rd kyverno.ResourceDescription) []*kyverno.ResourceSpec {
 	return specs
 }
 
+func generateTriggers(client client.Interface, rule kyverno.Rule, log logr.Logger) []*unstructured.Unstructured {
+	list := &unstructured.UnstructuredList{}
+
+	kinds := fetchUniqueKinds(rule)
+
+	for _, kind := range kinds {
+		mlist, err := client.ListResource("", kind, "", rule.MatchResources.Selector)
+		if err != nil {
+			log.Error(err, "failed to list matched resource")
+		}
+		list.Items = append(list.Items, mlist.Items...)
+	}
+
+	for _, any := range rule.MatchResources.Any {
+		for _, kind := range kinds {
+			anyList, err := client.ListResource("", kind, "", any.Selector)
+			if err != nil {
+				log.Error(err, "failed to list any matched resource")
+			}
+			list.Items = append(list.Items, anyList.Items...)
+		}
+	}
+
+	var kindlist []string
+	for _, all := range rule.MatchResources.All {
+		kindlist = append(kindlist, all.Kinds...)
+	}
+
+	if isMatchResourcesAllValid(kindlist) {
+		for _, all := range rule.MatchResources.All {
+			for _, kind := range kinds {
+				allList, err := client.ListResource("", kind, "", all.Selector)
+				if err != nil {
+					log.Error(err, "failed to list all matched resource")
+				}
+				list.Items = append(list.Items, allList.Items...)
+			}
+		}
+	}
+	// construct unique resource list
+	return constructUniquelist(list.Items)
```

Comment:
Let's build the unique kind list and use that for client listing. This way we will get a unique list of objects and don't need to  `constructUniquelist` again. https://github.com/kyverno/kyverno/pull/3819/files#diff-e59655168071c435149d448cc32f69e627b318dcb62a2438497c6b9e3e40baadR46-R51